src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java

```diff
@@ -48,234 +73,275 @@
-   * @param watchOption key option
-   * @param callback call back
-   * @return CompletableFuture watcher
-   */
   @Override
-  public CompletableFuture<Watcher> watch(ByteSequence key, WatchOption watchOption,
-      WatchCallback callback) {
-    WatchRequest request = optionToWatchCreateRequest(Util.byteStringFromByteSequence(key),
-        watchOption);
-    WatcherImpl watcher = new WatcherImpl(key, watchOption, callback);
-    CompletableFuture<Watcher> waitFuture = new CompletableFuture<Watcher>();
-    this.pendingCreateWatchers.add(new Pair<>(watcher, waitFuture));
-    getRequestStream().onNext(request);
-    return waitFuture;
+
+  public Watcher watch(ByteSequence key) {
+    return this.watch(key, WatchOption.DEFAULT);
   }
 
-  /**
-   * Cancel the watch task with the watcher, the onCanceled will be called after successfully
-   * canceled.
-   *
-   * @param id the watcher to be canceled
-   */
-  protected CompletableFuture<Boolean> cancelWatch(long id) {
-    WatcherImpl temp = watchers.get(id);
-    CompletableFuture<Boolean> completableFuture = null;
-    if (temp != null) {
-      synchronized (temp) {
-        if (this.watchers.containsKey(temp.getWatchID())) {
-          this.watchers.remove(temp.getWatchID());
-          completableFuture = new CompletableFuture<>();
-          this.pendingCancelFutures.put(id, completableFuture);
-        }
-      }
+  @Override
+  public synchronized Watcher watch(ByteSequence key, WatchOption watchOption) {
+    Preconditions.checkState(!isClosed(), "Watch client has been closed");
+
+    WatcherImpl watcher = new WatcherImpl(key, watchOption);
+    this.pendingWatchers.add(watcher);
+
+    if (this.pendingWatchers.size() == 1) {
+      // head of the queue send watchCreate request.
+      WatchRequest request = this.toWatchCreateRequest(watcher);
+      this.getGrpcWatchStreamObserver().onNext(request);
     }
 
-    WatchCancelRequest cancelRequest = WatchCancelRequest.newBuilder().setWatchId(id).build();
-    WatchRequest request = WatchRequest.newBuilder().setCancelRequest(cancelRequest).build();
-    getRequestStream().onNext(request);
-    return completableFuture;
+    return watcher;
   }
 
-  /**
-   * empty the old request stream, watchers and resume the old watchers empty the
-   * pendingCancelFutures as there is no need to cancel, the old request stream has been dead.
-   */
-  private synchronized void resume() {
-    this.requestStream = null;
-    WatcherImpl[] resumeWatchers = watchers.values().toArray(new WatcherImpl[watchers.size()]);
+  @Override
+  public synchronized void close() {
+    Preconditions.checkState(!isClosed(), "Watch client has been closed");
+
+    this.setClosed();
+    this.notifyWatchers(EtcdExceptionFactory.newEtcdException("Watch client has been closed"));
+    this.closeGrpcWatchStreamObserver();
+    this.executor.shutdownNow();
+    this.scheduledExecutorService.shutdownNow();
+  }
+
+  private void notifyWatchers(EtcdException e) {
```

Comment:
the log message is not accurate. the client should notify the watchers with a exception but not closing it. user is responsible for closing watcher.